okhttp/src/main/java/okhttp3/internal/concurrent/TaskQueue.kt

```diff
@@ -101,25 +100,42 @@ class TaskQueue internal constructor(
     }, delayNanos)
   }
 
-  /** Returns true if this queue became idle before the timeout elapsed. */
-  fun awaitIdle(delayNanos: Long): Boolean {
-    val latch = CountDownLatch(1)
+  /** Returns a latch that reaches 0 when the queue is next idle. */
+  fun idleLatch(): CountDownLatch {
+    synchronized(taskRunner) {
+      // If the queue is already idle, that's easy.
+      if (activeTask == null && futureTasks.isEmpty()) {
+        return CountDownLatch(0)
+      }
 
-    val task = object : Task("OkHttp awaitIdle", cancelable = false) {
-      override fun runOnce(): Long {
-        latch.countDown()
-        return -1L
+      // If there's an existing AwaitIdleTask, use it. This is necessary when the executor is
+      // shutdown but still busy as we can't enqueue in that case.
+      val existingTask = activeTask
+      if (existingTask is AwaitIdleTask) {
+        return existingTask.latch
+      }
+      for (futureTask in futureTasks) {
```

Comment:
Not sure about practically, but do you need to check anything but the last task here?  If it's non empty and has other tasks, seems by definition it's not idle.

Reply:
I think at the edges I’m conflating two concepts:
 * idle: nothing running, nothing scheduled
 * flush: complete all currently-enqueued tasks

In practice they’re the same effect for us, but I’ve implemented flush and labeled it idle. Because true idle is more work to implement for academic benefit I’m going to leave it as-is for now. Unsatisfying!

Reply:
One other wrench here . . . idle is distinct from shutdown. For example in our test suite we enforce idleness between tests, but share the same task runner between tests.

Reply:
Hopefully this is still a remaining cause of when our tests fail they tend to be flaky in batches, and we can fix a few remaining issues and have it rock solid.

Overall - trending ok https://circleci.com/build-insights/gh/square/okhttp/master

But big room for improvements once the base is solid.